routing/features_road_graph.cpp

```diff
@@ -294,31 +316,27 @@ IRoadGraph::RoadInfo const & FeaturesRoadGraph::GetCachedRoadInfo(FeatureID cons
 
   // ft must be set
   ASSERT_EQUAL(featureId, ft.GetID(), ());
-
-  ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
-
-  ri.m_bidirectional = !IsOneWay(ft);
-  ri.m_speedKMPH = speedKMPH;
-  ft.SwapPoints(ri.m_points);
-
-  LockFeatureMwm(featureId);
-
+  ExtractRoadInfo(featureId, ft, speedKMPH, ri);
   return ri;
 }
 
-void FeaturesRoadGraph::LockFeatureMwm(FeatureID const & featureId) const
+FeaturesRoadGraph::Value const & FeaturesRoadGraph::LockFeatureMwm(FeatureID const & featureId) const
 {
   MwmSet::MwmId mwmId = featureId.m_mwmId;
   ASSERT(mwmId.IsAlive(), ());
 
   auto const itr = m_mwmLocks.find(mwmId);
   if (itr != m_mwmLocks.end())
-    return;
+    return itr->second;
 
   MwmSet::MwmHandle mwmHandle = m_index.GetMwmHandleById(mwmId);
   ASSERT(mwmHandle.IsAlive(), ());
 
-  m_mwmLocks.insert(make_pair(move(mwmId), move(mwmHandle)));
-}
+  MwmValue * mwmValue = nullptr;
+  if (mwmHandle.IsAlive())
+    mwmValue = mwmHandle.GetValue<MwmValue>();
 
+  Value value(move(mwmHandle), mwmValue);
```

Comment:
I rewritten LockFeatureMwm and callee code a little. But I do not think it's the best solution. Anyway we still can have "!mwmHandle.IsAlive()" mwmHandle(s) in m_mwmLocks (after ASSERT). 

LockFeatureMwm is called only from ExtractRoadInfo methods. Is it possible to get an invalid mwmHandle while building route? I guess yes. If so, I cannot write CHECK(mwmHandle.IsAlive()) instead of ASSERT.